### PR TITLE
Introduce LibraryQualityEvaluator	

### PR DESCRIPTION
**Pull request by leogrim**
@ymatsuda @stkem @yingjieMiao @andrewconner

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  private val mu = Math.log(mode + Math.pow(sigma, 2))
+  private val maxProbability = logNormalDensity(mu, sigma, mode)
+
+  @inline private def logNormalDensity(mu: Double, sigma: Double, x: Double): Double = {
+    Math.exp(-Math.pow(Math.log(x) - mu, 2) / (2 * Math.pow(sigma, 2))) / (x * sigma * Math.sqrt(2 * Math.PI))
+  }
+
+  private val LowQualityLibraryNamesRe = "(?i)(test|delicious|bookmark|pocket|kippt|asdf|pinboard|import|instapaper)".r
+  def isPoorlyNamed(name: String): Boolean = {
+    name.size <= 2 || LowQualityLibraryNamesRe.findFirstIn(name).isDefined || name.toLowerCase.split("\\s+").exists(Profanity.all.contains)
```
Comment (eishay) [inline kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala]: We have few legit libraries named AI, V6, UI, ML, AI, IT, R, UX, VC, Pi, JS, C and more. There may be many false positives here.

**Comment by eishay** [comment]
great

File: kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala]: Took that assertion from Curator actually. I'll remove it. @joshm1

File: kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  private val mode = optimalLibrarySize / activeShards.all.size // assuming uniform distribution of a library's keeps across all shards
+  private val sigma = 1.0 // eyeballed for reasonable decay
+  private val mu = Math.log(mode + Math.pow(sigma, 2))
+  private val maxProbability = logNormalDensity(mu, sigma, mode)
+
+  @inline private def logNormalDensity(mu: Double, sigma: Double, x: Double): Double = {
+    Math.exp(-Math.pow(Math.log(x) - mu, 2) / (2 * Math.pow(sigma, 2))) / (x * sigma * Math.sqrt(2 * Math.PI))
+  }
+
+  private val LowQualityLibraryNamesRe = "(?i)(test|delicious|bookmark|pocket|kippt|asdf|pinboard|import|instapaper)".r
```
Comment (andrewconner) [inline kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala:38]: Hm, should we include `twitter` and `tweet`? Right now they seem pretty catch-all

Comment (stkem) [inline kifi-backend/modules/search/app/com/keepit/search/engine/LibraryQualityEvaluator.scala:38]: +1, at least for now

**Pull request merged**